feat: Add Cloud Function execution on Parse Object in data browser (#2409)
parse-community/parse-dashboard: README.md

  - [Other Configuration Options](#other-configuration-options)
    - [Prevent columns sorting](#prevent-columns-sorting)
    - [Custom order in the filter popup](#custom-order-in-the-filter-popup)

>     - [Custom Cloud Code scripts](#custom-cloud-code-scripts)

Reviewer: ```suggestion
    - [Scripts](#scripts)
```

          {
            "name": "email",
            "filterSortToTop": true
          }
        ]
      }
    }
]
```

> ### Custom Cloud Code scripts
> 
> You can specify custom cloud code scripts with the `scripts` option:

Reviewer: I'd name this feature more general just "Scripts". Maybe in the future we add a parameter `url` so that instead of a Cloud Function any RESTful API can be invoked with the Parse Object as payload.

```suggestion
### Scripts

You can specify scripts to execute Cloud Functions with the `scripts` option:
```

```json
"apps": [
  {
    "scripts": [
      {
        "title": "Delete Account",
        "classes": ["_User"],
        "cloudCodeFunction": "deleteAccount"
      }
    ]
  }
]
```

> Next, define a cloud function within the connected Parse Server.

Reviewer: ```suggestion
Next, define a Cloud Function within the connected Parse Server:
```

> **Parse Server 2.1.4 > 4.4.0**
> 
> ```js
> Parse.Cloud.define('deleteAccount', async (req) => {
>   if (!req.master || !req.params.object) {
>     throw 'Unauthorized';
>   }
>   req.params.object = Parse.Object.fromJSON(req.params.object);
>   req.params.object.set('deleted', true);
>   await req.params.object.save(null, {useMasterKey: true});
> });
> ```
> 
> **Parse Server 4.4.0 > 6.2.0**
> 
> ```js
> Parse.Cloud.define('deleteAccount', async (req) => {
>   req.params.object = Parse.Object.fromJSON(req.params.object);
>   req.params.object.set('deleted', true);
>   await req.params.object.save(null, {useMasterKey: true});
> }, {
>   requireMaster: true
> });
> ```
> 
> **Parse Server 6.2.0+**
> 
> ```js
> Parse.Cloud.define('deleteAccount', async (req) => {
>   req.params.object.set('deleted', true);
>   await req.params.object.save(null, {useMasterKey: true});
> }, {
>   requireMaster: true
> });
> ```

Reviewer: Not sure how the GitHub MD parser likes the `&gt;`and `&lt;`...

````suggestion
```js
Parse.Cloud.define('deleteAccount', async (req) => {
  req.params.object.set('deleted', true);
  await req.params.object.save(null, {useMasterKey: true});
}, {
  requireMaster: true
});
```

For older versions of Parse Server:

<details>
<summary>Parse Server &gt;=4.4.0 &lt;6.2.0</summary>

```js
Parse.Cloud.define('deleteAccount', async (req) => {
  req.params.object = Parse.Object.fromJSON(req.params.object);
  req.params.object.set('deleted', true);
  await req.params.object.save(null, {useMasterKey: true});
}, {
  requireMaster: true
});
```

</details>

<details>
<summary>Parse Server &gt;=2.1.4 &lt;4.4.0</summary>

```js
Parse.Cloud.define('deleteAccount', async (req) => {
  if (!req.master || !req.params.object) {
    throw 'Unauthorized';
  }
  req.params.object = Parse.Object.fromJSON(req.params.object);
  req.params.object.set('deleted', true);
  await req.params.object.save(null, {useMasterKey: true});
});
```

</details>
````

You can conveniently create a filter definition without having to write it by hand by first saving a filter in the data browser, then exporting the filter definition under *App Settings > Export Class Preferences*.

### Scripts

You can specify scripts to execute Cloud Functions with the `scripts` option:

```json
"apps": [
  {
    "scripts": [
      {
        "title": "Delete Account",
        "classes": ["_User"],
        "cloudCodeFunction": "deleteAccount"
      }
    ]
  }
]
```

Next, define a Cloud Function within the connected Parse Server:

```js
Parse.Cloud.define('deleteAccount', async (req) => {
  req.params.object.set('deleted', true);
  await req.params.object.save(null, {useMasterKey: true});
}, {
  requireMaster: true
});
```

For older versions of Parse Server:

<details>
<summary>Parse Server &gt;=4.4.0 &lt;6.2.0</summary>

```js
Parse.Cloud.define('deleteAccount', async (req) => {

>   req.params.object = Parse.Object.fromJSON(req.params.object);

Reviewer: @dblythy we've always been mentioning pointers which cause the issue in Parse Server. I though of it like pointers within a Parse Object, that point to another object. But from the example it seems that the object itself needs to be transformed from POJO to a `Parse.Object` instance. What changes are necessary in the README for this to get merged?

You can conveniently create a filter definition without having to write it by hand by first saving a filter in the data browser, then exporting the filter definition under *App Settings > Export Class Preferences*.

### Scripts

You can specify scripts to execute Cloud Functions with the `scripts` option:

```json
"apps": [
  {
    "scripts": [
      {
        "title": "Delete Account",
        "classes": ["_User"],
        "cloudCodeFunction": "deleteAccount"
      }
    ]
  }
]
```

Next, define the Cloud Function in Parse Server that will be called. The object that has been selected in the data browser will be made available as a request parameter:

```js
Parse.Cloud.define('deleteAccount', async (req) => {
  req.params.object.set('deleted', true);
  await req.params.object.save(null, {useMasterKey: true});
}, {
  requireMaster: true
});
```

> ⚠️ Depending on your Parse Server version you may need to set the Parse Server option `encodeParseObjectInCloudFunction` to `true` so that the selected object in the data browser is made available in the Cloud Function as an instance of `Parse.Object`. If the option is not set, is set to `false`, or you are using an older version of Parse Server, the object needs to be manually converted from a JSON object to a `Parse.Object` instance with `req.params.object = Parse.Object.fromJSON(req.params.object);`, before you can call any `Parse.Object` properties and methods on it.

Reviewer: ```suggestion
⚠️ Depending on your Parse Server version you may need to set the Parse Server option `encodeParseObjectInCloudFunction` to `true` so that the selected object in the data browser is made available in the Cloud Function as an instance of `Parse.Object`. If the option is not set, is set to `false`, or you are using an older version of Parse Server, the object is made available as a plain JavaScript object and needs to be converted from a JSON object to a `Parse.Object` instance with `req.params.object = Parse.Object.fromJSON(req.params.object);`, before you can call any `Parse.Object` properties and methods on it.
```

You can conveniently create a filter definition without having to write it by hand by first saving a filter in the data browser, then exporting the filter definition under *App Settings > Export Class Preferences*.

### Scripts

You can specify scripts to execute Cloud Functions with the `scripts` option:

```json
"apps": [
  {
    "scripts": [
      {
        "title": "Delete Account",
        "classes": ["_User"],
        "cloudCodeFunction": "deleteAccount"
      }
    ]
  }
]
```

> Next, define a Cloud Function within the connected Parse Server, and set `encodePointerInCloudFunction` in the server configuration to `true`:

Reviewer: ```suggestion
Next, define a Cloud Function within the connected Parse Server, and set `encodeParseObjectInCloudFunction` in the server configuration to `true`:
```

```js
Parse.Cloud.define('deleteAccount', async (req) => {
  req.params.object.set('deleted', true);
  await req.params.object.save(null, {useMasterKey: true});
}, {
  requireMaster: true
});
```

For older versions of Parse Server:

<details>

> <summary>encodePointerInCloudFunction set to false</summary>

Reviewer: ```suggestion
<summary> encodeParseObjectInCloudFunction set to false</summary>
```

> Next, define a Cloud Function within the connected Parse Server, and set `encodeParseObjectInCloudFunction` in the server configuration to `true`:
> 
> ```js
> Parse.Cloud.define('deleteAccount', async (req) => {
>   req.params.object.set('deleted', true);
>   await req.params.object.save(null, {useMasterKey: true});
> }, {
>   requireMaster: true
> });
> ```
> 
> For older versions of Parse Server:
> 
> <details>
> <summary> encodeParseObjectInCloudFunction set to false</summary>
> 
> ```js
> Parse.Cloud.define('deleteAccount', async (req) => {
>   req.params.object = Parse.Object.fromJSON(req.params.object);
>   req.params.object.set('deleted', true);
>   await req.params.object.save(null, {useMasterKey: true});
> }, {
>   requireMaster: true
> });
> ```
> 
> </details>

Reviewer: ````suggestion
Next, define the Cloud Function in Parse Server that will be called. The object that has been selected in the data browser will be made available as a request parameter:

```js
Parse.Cloud.define('deleteAccount', async (req) => {
  req.params.object.set('deleted', true);
  await req.params.object.save(null, {useMasterKey: true});
}, {
  requireMaster: true
});
```

⚠️ Depending on your Parse Server version you may need to set the Parse Server option `encodeParseObjectInCloudFunction` to `true` so that the selected object in the data browser is made available in the Cloud Function as an instance of `Parse.Object`. If the option is not set, is set to `false`, or you are using an older version of Parse Server, the object needs to be manually converted from a JSON object to a `Parse.Object` instance with `req.params.object = Parse.Object.fromJSON(req.params.object);`, before you can call any `Parse.Object` properties and methods on it.

For older versions of Parse Server:

````
